### ISSUE TITLE
GLTFimporter prevent duplicates of same image when used in multiple materials

### ISSUE THREAD

**Issue by optimus007** [mrdoob/three.js#21559]
NOTE :Old title was  'GLTFexporter prevent exporting duplicates of same image when used in multiple materials'

**Is your feature request related to a problem? Please describe.**

same images used in multiple materials exports as duplicates leading to increase in file size

**Describe the solution you'd like**

If same textureLoader object is used then have the exporter save the image once & have the materials refer to this single image 

**Additional context**

Also applicable for other exporters like USDZ

**Comment by donmccurdy** [comment]
There is a cache intended to prevent this, so I have to assume it prevents some duplicates, would you be able to share details to reproduce the case where it does not?

https://github.com/mrdoob/three.js/blob/71759eb24a2fdeee3de21d6380dcc7d1408b1402/examples/js/exporters/GLTFExporter.js#L924-L930

**Comment by optimus007** [comment]
@donmccurdy 
Made a simple version to show the issue 
i'm currently using r125

This is how i created the glb in blender 2.92 . image is same in both and one material has color multiplied
<img width="1507" alt="input" src="https://user-images.githubusercontent.com/6885294/113381017-7a89d780-939b-11eb-8a1f-9bd112b3532d.png">


This is the glb generated by gltfexporter. opened in blender 
<img width="1522" alt="output" src="https://user-images.githubusercontent.com/6885294/113381093-af962a00-939b-11eb-936c-180a990b2902.png">

the glb
[glb.zip](https://github.com/mrdoob/three.js/files/6247747/glb.zip)

**Comment by Mugen87** [comment]
> um...after looking into it more ,This might be due to the gltfImporter ... both materials seems to have images with different uuid

I've investigated this issue today and indeed the problem is that the loader already creates two textures with different image/image bitmap objects. This makes it of course more or less impossible for the exporter to avoid duplicate image data.

Setting `THREE.Cache.enabled` to `true` does not solve the issue.

**Comment by Mugen87** [comment]
When an image of a glTF asset is defined by a reference to a buffer view, the loader creates a respective URI via `createObjectURL()`. 

https://github.com/mrdoob/three.js/blob/979d8f284cffe9bbf9d618e67462b69e5e9d430d/examples/jsm/loaders/GLTFLoader.js#L2561

However, this happens right now every time when a texture refers to this image. I guess the loader should create the URI only once and then safe it for further textures (e.g. by assigning it to `source.uri`).

**Comment by optimus007** [comment]
should i create a new issue & close this one ?

**Comment by Mugen87** [comment]
Please no. Let's use this one.

**Comment by Mugen87** [comment]
I guess solving #17766 would make it easier to handle this use case in the loader and thus in the exporter.

**Comment by donmccurdy** [comment]
The file contains these texture and image entries:

```js
"textures": [
  {
    "sampler": 0,
    "source": 0
  },
  {
    "sampler": 0,
    "source": 0
  }
],
"images": [
  {
    "mimeType": "image/jpeg",
    "name": "REDgrad",
    "uri": "REDgrad.jpg"
  }
],
```

That's redundant, and the duplicate texture entries, pointing to the same image, are what caused the issue here. @optimus007 if you would you be able to share the original `.blend` I would like to report this to the Blender exporter and get it fixed there if possible.

In the meantime #21821 should fix this and another similar issue.

**Comment by optimus007** [comment]
@donmccurdy sure 

blender 2.92 zip contains glb and blend with packed image
[2Materials1Image.zip](https://github.com/mrdoob/three.js/files/6470524/2Materials1Image.zip)

**Comment by optimus007** [comment]
@donmccurdy is there a temporary fix i can perform in the meantime ?
i have these human models (made by someone else) which uses single image set but shared across multiple materials
eg skin,shirt,pant,hair, etc but there's just one packed pbr texture set

**Comment by donmccurdy** [comment]
#21821 will be in the next release of three.js.

In the meantime, running the model through `gltf-transform cp input.glb output.gltf` (to fix the extra texture entry) and loading it as .gltf rather than .glb should work around the issue.
